BSIP31-34 market engine improvements

abitmore opened the pull request:
This is a subset of changes on the market engine, cherry-picked from #641, including changes related to #338 #343 #453 #606 #625 #649, aka
* [BSIP31: Update Short Position's Margin Call Price After Partially Called Or Settled](https://github.com/bitshares/bsips/blob/master/bsip-0031.md), 
* [BSIP32: Always Match Orders At Maker Price](https://github.com/bitshares/bsips/blob/master/bsip-0032.md), 
* [BSIP33: Maker Orders With Better Prices Take Precedence](https://github.com/bitshares/bsips/blob/master/bsip-0033.md), 
* [BSIP34: Always Trigger Margin Call When Call Price Above Or At Price Feed](https://github.com/bitshares/bsips/blob/master/bsip-0034.md).

Please review.

abitmore commented on libraries/chain/db_market.cpp:
```diff
@@ -711,27 +858,35 @@ bool database::check_call_orders(const asset_object& mia, bool enable_black_swan
        {
-          wlog( "Cull_small issue occurred at block #${block}", ("block",head_block_num()) );
-          limit_itr = next_limit_itr;
+          if( filled_limit ) ++limit_itr;
```
Note: the code block here, including `++limit_itr` here and below, has been refactored in a later commit, we'll need to review again.

pmconrad commented on libraries/chain/db_market.cpp:
```diff
@@ -643,8 +800,11 @@ bool database::check_call_orders(const asset_object& mia, bool enable_black_swan
     bool filled_limit = false;
```
Please add an FC_ASSERT to check_call_orders that ensures that for_new_limit_order is only true before HF 338 / 625 (see my remark on check_for_blackswan below).

abitmore replied:
Similar as commented above, actually we need to carefully review all `assert()`s and `FC_ASSERT()`s as mentioned in https://github.com/bitshares/bitshares-core/issues/679 and https://github.com/bitshares/bitshares-core/issues/511#issuecomment-366802721, also be careful when adding new `FC_ASSERT()`s. Actually I want to get rid of all `FC_ASSERT()`s or catch them properly when processing post-transaction jobs in `_apply_block()`.

For this one, if we add a `FC_ASSERT()` here then caused a chain halt, probably would be caused by `clear_expired_feeds` or FBA buyback.

pmconrad replied:
After HF 388, check_call_orders / check_for_blackswan only operate correctly if the maker flag is false. Adding the suggested FC_ASSERT would enforce that this condition holds even if a future change mistakenly calls check_call_orders with the flag set to true.

General question (not rhetorical): what is worse, a chain halt or a chain operating with undefined and/or wrong behaviour?

abitmore replied:
IMHO at most times a chain halt is worse, but negative impact of undefined/wrong behavior is unbounded and sometimes is even worse. So I don't have a clear answer. I slightly tend to avoid chain halt.

pmconrad replied:
I think the opposite, especially because wrong behaviour can go unnoticed for a long time.

We need more opinions. :-) @xeroc @oxarbitrage  @ryanRfox

pmconrad commented on libraries/chain/db_maint.cpp:
```diff
@@ -766,6 +766,40 @@ void database::process_bids( const asset_bitasset_data_object& bad )
+      }
+      if( !abd || abd->is_prediction_market ) // nothing to do with PM's; check !abd just to be safe
+         continue;
+      db.modify( call_obj, [&]( call_order_object& call ) {
```
Please capture only abd

abitmore replied:
Done.

abitmore commented on libraries/chain/db_update.cpp:
```diff
@@ -217,38 +231,35 @@ bool database::check_for_blackswan( const asset_object& mia, bool enable_black_s
-       assert( settle_price.base.asset_id == limit_itr->sell_price.base.asset_id );
-       highest = std::max( limit_itr->sell_price, settle_price );
+       FC_ASSERT( highest.base.asset_id == limit_itr->sell_price.base.asset_id );
+       highest = std::max( limit_itr->sell_price, highest );
```
In short, this change is correct.

abitmore commented on libraries/chain/db_market.cpp:
```diff
@@ -364,18 +364,137 @@ bool database::apply_order(const limit_order_object& new_order_object, bool allo
-int database::match( const limit_order_object& usd, const OrderType& core, const price& match_price )
+int database::match( const limit_order_object& usd, const limit_order_object& core, const price& match_price )
 {
    assert( usd.sell_price.quote.asset_id == core.sell_price.base.asset_id );
```
Actually we need to carefully review all `assert()`s and `FC_ASSERT()`s as mentioned in https://github.com/bitshares/bitshares-core/issues/679 and https://github.com/bitshares/bitshares-core/issues/511#issuecomment-366802721.

For this one, if it would cause a chain halt, most probably would be caused by FBA buyback.

abitmore commented on libraries/chain/db_maint.cpp:
```diff
@@ -766,6 +766,40 @@ void database::process_bids( const asset_bitasset_data_object& bad )
+      const asset_object& a = *itr;
+      ++itr;
+      // be here, next_maintenance_time should have been updated already
+      db.check_call_orders( a, true, false ); // allow black swan, and call orders are taker
```
Whether to skip over PM's here, the only logic difference might be the `check_black_swan()` call in `check_call_orders()`.

Since we didn't fix https://github.com/bitshares/bitshares-core/issues/460 in this release (no BSIP for it so far), PM's will still suffer black swan events when there is a price feed change or matching logic change.

Although we didn't update `call_price`s of PM's, but we changed logic in `check_black_swan()` to check with MSSP, which is possible to trigger a black swan event. So the outcome is different if skip over PM's here.

I'd suggest we trigger black swan events here (if there would be some) to be consistent (rather than triggering on a new but same price feed).

abitmore replied:
Perhaps keep the old logic for PM's, don't trigger black swan event when there is no matching limit order?

abitmore replied:
My opinion on #460 is here: https://github.com/bitshares/bitshares-core/issues/460#issuecomment-364520866. Aka not really a bug and not a quick fix to me.

So my recommendation in this release is to let PMs (still) have same behavior (on black swan events) as other MPAs, that said, we changed behavior of other MPAs, which will affect PMs as well.

pmconrad commented:
Please add/modify a test to cover this code path: https://sonarcloud.io/component?id=BitShares-develop%3Adevelop%3Alibraries%2Fchain%2Fdb_market.cpp&line=561

(Create a large limit order that is only partially filled by a margin call.)

abitmore commented:
@pmconrad done changes for this:
> (Create a large limit order that is only partially filled by a margin call.)

abitmore commented:
@pmconrad I've changed the `assert()`s to `FC_ASSERT()`s and added new `FC_ASSERT()` as requested. Please check again. WRT the PM black swan issue (#460), my opinion is we don't fix it this time.
